keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -366,61 +368,100 @@ The `kubernetes.io/ingress.class` annotation is required for selecting between
 multiple Ingress providers. As support for this annotation is universal, this
 concept should be promoted to an actual field.
 
-#### Ingress class proposal
+#### Ingress class field
 
-Promoting the annotation as it is currently defined as an opaque string is the
-most direct path but precludes any future enhancements to the concept.
-
-An alternative is to create a new resource `IngressClass` to take its place.
-This resource will serve a couple of purposes:
-
-- Define the set of valid classes available to the user. Gives operators control
-  over allowed classes.
-- Allow us to evolve the API to express concepts such a levels of service
-  associated with a given Ingress controller.
-
-Add a field to `ingress.spec`:
+Although promoting the annotation as it is currently defined as an opaque string
+is the most direct path, that precludes any future enhancements to the concept.
+With that in mind, we propose creating a new `Class` field in `IngressSpec` to
+take the place of the existing annotation. This new field will be immutable. To
+ensure that this can be safely round tripped between API versions, this new
+field will also be added to previous API versions.
 
 ```golang
 type IngressSpec struct {
   ...
-  // Class is the name of the IngressClass cluster resource. This defines
-  // which controller(s) will implement the resource.
-  Class string
+  // Class is the name of the IngressClass cluster resource. This defines which
+  // controller(s) will implement the resource.
+  // +optional
+  Class *string
   ...
 }
+```
+
+##### Interoperability with previous annotation
+
+The `kubernetes.io/ingress.class` annotation will be separate from the new Class
+field. As the annotation was never formally defined or validated, we can not
+safely convert the value of this annotation to the new Class field. Use of this
+annotation will be considered formally deprecated with the v1 Ingress release.
+
+When both the class field and annotation are set, the annotation will take
+priority. The controller MAY emit a warning event if the user sets conflicting
+(different) values for the annotation and field.
+
+To ensure backwards compatibility, Ingresses may be created without a class
+field or annotation. Although this is not a recommended state, it must still be
+supported by the API. Implementations of this API may choose to ignore Ingresses
+without a class specified. In certain cases, such as when an Ingress class is
+marked as default, it may make sense for Ingress implementations to implement
+Ingresses that do not have a class specified.
 
-...
+When new Ingresses are created, if both the class field and annotation are set,
```

Comment:
How about on update?  If we can't validate this on update (as we discussed) then we should say so.